scalableminds/webknossos
# Allow multiple contributors per annotation

It is now possible to allow other users, who may see your annotation, to also edit it.

### URL of deployed dev instance (used for testing):
- https://annotationcontributors.webknossos.xyz

### Backend API Changes
 - annotation json now includes new field `contributors: List[User]`, listing the users who pushed update actions, but who are not the owner
 - update actions in `updateActionLog` now contain new optional field `authorId: String` containing a user ID. Legacy update actions without this field should be considered to be authored by the annotation owner.
 - new route `PATCH /api/annotations/:typ/:id/othersMayEdit?othersMayEdit=true` allows to set this permission
 - users who can see an annotation with `othersMayEdit==true` will get `allowUpdate: true` in the annotation json
 - annotation json now includes new field `othersMayEdit: boolean` indicating whether the above option has been set for the annotation

### Steps to test:
- ensure you have two separate users set up (one logged-in in incognito tab)
- create a new annotation
- open the share modal and assign a orga team to the "allowed teams"
- save
- open the annotation link for user B and check that the annotation is in read-only mode
- open the share modal again for user A and change the "allow others" setting to true
- reload the annotation for user B --> annotation should be editable
- trace a bit in window B and save
- wait a bit in window A --> a toast should appear warning about conflicts

### Issues:
- fixes #6227 

------
